main/core/src/eval/Evaluator.scala

```diff
@@ -362,15 +382,302 @@ case class Evaluator(home: os.Path,
+        work = newWork
+        inProgress = inProgress ++ newInProgress
+
+        evalLog.debug(s"Search for ${newInProgress.size} new dep-free tasks took ${System.currentTimeMillis() - scheduleStart} msec")
+
+        if (!newInProgress.isEmpty) {
+          evalLog.debug(s"Scheduling ${newInProgress.size} new tasks (issuer: ${issuer}): ${newInProgress.map(t => printTerm(t._1)).mkString(", ")}")
+
+          // schedule for parallel execution
+          newInProgress.zipWithIndex.foreach {
+            case (curWork @ (terminal, group), index) =>
+
+              //              val missingDependencies = group.indexed.flatMap(_.inputs).filter(t => !results.contains(t) && !group.contains(t))
+              //              if (!missingDependencies.isEmpty) {
+              //                evalLog.error(s"Missing resolved dependencies for terminal group: ${printTerm(terminal)}\n  ${missingDependencies.mkString(",\n  ")}")
+              //              }
+
+              val workerFut: java.util.concurrent.Future[FutureResult] = completionService.submit { () =>
+                if(failFast && someTaskFailed) {
+                  // we do not start this tasks but instead return with aborted result
+                  val newResults = group.map { task =>
+                    task -> Aborted
+                  }
+
+                  evalLog.debug(s"Skipped evaluation (because of earlier failures): ${printTerm(terminal)}")
+                  FutureResult(curWork, 0, Evaluated(newResults.toMap, Seq(), false))
+
+                } else {
+
+                  val counterMsg = nextCounterMsg()
+
+                  evalLog.debug(s"Start evaluation [${counterMsg}]: ${printTerm(terminal)}")
+                  val startTime = System.currentTimeMillis()
+
+                  val res = evaluateGroupCached(
+                    terminal,
+                    group,
+                    results,
+                    counterMsg,
+                    reporter,
+                    testReporter,
+                    logger
+                  )
+
+                  val endTime = System.currentTimeMillis()
+                  evalLog.debug(s"Finished evaluation [${counterMsg}]: ${printTerm(terminal)}")
+
+                  FutureResult(curWork, (endTime - startTime).toInt, res)
+                }
+              }
+
+              evalLog.debug(s"New future: ${workerFut} [${index + 1}/${newInProgress.size}] for task: ${printTerm(terminal)}")
+              futures = futures ++ List(workerFut -> curWork)
+          }
+        } else {
+          evalLog.debug(s"No new tasks to schedule (issuer: ${issuer})")
+        }
+      }
+
+      scheduleWork("initial request")
+
+      // Work queue management
+      // wait for finished jobs and schedule more work, if possible
+      while (futures.size > 0) {
+        evalLog.debug(s"Waiting for next future completion of ${executorService}")
+        val compFuture: Future[FutureResult] = completionService.take()
```

Comment:
This loop driving execution and scheduling of futures seems pretty complicated. Could it be simpler to use Scala `Future`s instead? Then we could construct the future pipeline using `Future.traverse`, and their management and execution would be entirely handled for us by the `ExecutionContext`. We could even use an `ExecutionContext` that wraps the local `ThreadPoolExecutor` to preserve that property of the PR implementation

Reply:
The idea behind the use of Java futures was that they can be cancelled. In fact, in SBuild I used Scala futures and you are right, that was a lot easier. But exception propagation wasn't really nice, IIRC. I'll think about it and do some experiments. It would mean a huge change, as most of this PR is essentially future handling. On the other side, I wanted to touch exception handling anyways.

Reply:
I think for now let's ignore cancellation. Mill itself doesn't support fine-grained cancellation even when single threaded - it just kills the process when you Ctrl-C - and other tools like SBT or Bazel have a similar behavior. Task-level cancellation might be something that's worth pioneering, but we don't need to pioneer it in this first implementation of parallelism.

I know it would be a huge change, but hopefully it'll end up with a much smaller amount of much simpler code, which would help ensure correctness and make future maintenance easier.

I could also take a crack at a future-based implementation on top of this PR if you don't have time.

Reply:
I will not find the time to make that change in the next 1-2 weeks. It think we should do it nonetheless, if it makes code easier to reason about. But we can also to it as a follow up PR, which might make a lot of sense, as we would have a good base to compare against.